File: 0023.合并K个排序链表/0023-合并K个排序链表.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
bool cmp(const ListNode* a, const ListNode* b){
    return a->val < b->val;
}
class Solution {
public:
    ListNode* mergeKLists(vector<ListNode*>& lists) {
        std::vector<ListNode *> value;
        for(int i = 0; i< lists.size(); ++i){
            ListNode* head = lists[i];
            while(head){
                value.push_back(head);
                head = head->next;
            }
        }
        
        std::sort(value.begin(),value.end(), cmp);
        if (value.size() == 0)
            return NULL;
        
        for(int i = 1; i<value.size(); ++i){
            value[i-1]->next = value[i];
        }
        value[value.size()-1]->next = NULL;
        return value[0];
        
    }
};

File: 0023.合并K个排序链表/0023-合并K个排序链表.py
# Definition for singly-linked list.
# class ListNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution(object):
    def mergeKLists(self, lists):
        """
        :type lists: List[ListNode]
        :rtype: ListNode
        """
        from heapq import *
        pq = []
        for i in range(len(lists)):
            if lists[i]:
                heappush(pq, (lists[i].val, i))
                lists[i] = lists[i].next
            
        dummy = ListNode(1)
        p = dummy
        while pq:
            val, idx = heappop(pq)
            p.next = ListNode(val)
            p = p.next
            if lists[idx]:
                heappush(pq, (lists[idx].val, idx))
                lists[idx] = lists[idx].next
        return dummy.next
            